sync-android/src/main/java/com/cloudant/sync/datastore/encryption/KeyManager.java

```diff
@@ -0,0 +1,217 @@
+                        (CDTENCRYPTION_KEYCHAIN_ENCRYPTIONKEY_SIZE);
+                byte[] salt = generateSecureRandomBytesWithLength
+                        (CDTENCRYPTION_KEYCHAIN_PBKDF2_SALT_SIZE);
+                byte[] iv = generateSecureRandomBytesWithLength
+                        (CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE);
+                SecretKey aesKey = pbkdf2DerivedKeyForPassword(password, salt,
+                        CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS,
+                        CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE);
+
+                byte[] encryptedDpkBytes = DKPEncryptionUtil.encryptAES(aesKey, iv, dpkBytes);
+
+                KeyData keyData = new KeyData(encryptedDpkBytes, salt, iv,
+                        CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS, CDTENCRYPTION_KEYCHAIN_VERSION);
+
+                if (this.storage.saveEncryptionKeyData(keyData)) {
+                    dpk = new EncryptionKey(dpkBytes);
+                }
+            }
+        } catch (Exception e) {
+            throw new DPKException("Failed to encrypt DPK", e);
+        }
+        return dpk;
+    }
+
+    /**
+     * Checks if the encrypted Data Protection Key (DPK) is inside the {@link SharedPreferences}.
+     *
+     * @return true if the encrypted DPK is inside the {@link SharedPreferences}, false otherwise
+     */
+    public boolean keyExists() {
+        return this.storage.encryptionKeyDataExists();
+    }
+
+    /**
+     * Clears security metadata from the {@link SharedPreferences}.
+     *
+     * @return Success (true) or failure (false)
+     */
+    public boolean clearKey() {
+        return this.storage.clearEncryptionKeyData();
+    }
+
+    // PRIVATE HELPER METHODS
+    private boolean validateEncryptionKeyData(KeyData data) {
+        if (data.getIv().length != CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE) {
+            LOGGER.warning("IV does not have the expected size: " +
+                    CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE + " bytes");
+            return false;
+        }
+        return true;
+    }
+
+    private SecretKey pbkdf2DerivedKeyForPassword(String password, byte[] salt, int
+            iterations, int length) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        if (length < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (password == null || password.length() < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (salt == null || salt.length < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (iterations < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        SecretKeyFactory pbkdf2Factory;
+        if (Build.VERSION.SDK_INT >= 19) {
+            // Use compatibility key factory required for backwards compatibility in API 19 and up.
```

Comment:
Changes requested in comment included in https://github.com/cloudant/sync-android/commit/dd276d263bfc9c5ddebb047cecd27230cb074f63